util/pd.go

```diff
@@ -0,0 +1,204 @@
+//
+// https://github.com/pingcap/tidb/tree/cc5e161ac06827589c4966674597c137cc9e809c/store/tikv/util/execdetails.go
+//
+
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+// pd request retry time when connection fail
+const (
+	pdRequestRetryTime = 10
+
+	storeMinResolvedTSPrefix = "pd/api/v1/min-resolved-ts"
+)
+
+// PdController manage get/update config from pd.
+type PdController struct {
+	addrs []string
+	cli   *http.Client
+}
+
+func NewPdController(
+	tlsConf *tls.Config,
+	pdAddrs []string,
+) *PdController {
+	for i, addr := range pdAddrs {
+		if !strings.HasPrefix(addr, "http") {
+			if tlsConf != nil {
+				addr = "https://" + addr
+			} else {
+				addr = "http://" + addr
+			}
+			pdAddrs[i] = addr
+		}
+	}
+
+	return &PdController{
+		addrs: pdAddrs,
+		cli:   httpClient(tlsConf),
+	}
+}
+
+// GetStoreMinResolvedTS get store-level min-resolved-ts from pd
+func (p *PdController) GetStoreMinResolvedTS(ctx context.Context, storeID uint64) (uint64, error) {
+	var err error
+	for _, addr := range p.addrs {
+		query := fmt.Sprintf("%s/store_id=%d", storeMinResolvedTSPrefix, storeID)
+		v, e := pdRequest(ctx, addr, query, p.cli, http.MethodGet, nil)
+		if e != nil {
+			log.Warn("failed to get min resolved ts", zap.String("addr", addr), zap.Error(e))
+			err = e
+			continue
+		}
+		log.Info("store min resolved ts", zap.String("resp", string(v)))
+		d := struct {
+			IsRealTime    bool   `json:"is_real_time,omitempty"`
+			MinResolvedTS uint64 `json:"min_resolved_ts"`
+		}{}
+		err = json.Unmarshal(v, &d)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		if !d.IsRealTime {
+			message := "store min resolved ts not enabled"
+			log.Error(message, zap.String("addr", addr))
+			return 0, errors.Trace(errors.New(message))
+		}
+		return d.MinResolvedTS, nil
+	}
+	return 0, errors.Trace(err)
+}
+
+// pdRequest is a func to send an HTTP to pd and return the result bytes.
+func pdRequest(
+	ctx context.Context,
+	addr string, prefix string,
+	cli *http.Client, method string, body io.Reader) ([]byte, error) {
+	_, respBody, err := pdRequestWithCode(ctx, addr, prefix, cli, method, body)
+	return respBody, err
+}
+
+func pdRequestWithCode(
```

Comment:
It seems that this function is unnecessary since the returned code has never been used.